File: java/src/com/framgia/Part1/getuserinfo/GetUserInfoBetter.java
package com.framgia.Part1.getuserinfo;

import java.util.Calendar;
import java.util.List;

public class GetUserInfoBetter {

    private final int REQUEST_TIMEOUT = 69;

    static class RestClient {
        /**
         * DEMO METHOD
         * This method make a request to server then receive a list of users
         */
        static List<User> execute(String method, String url, int timeout) {
            return null;
        }
    }

    // get user from the internet
    private List<User> fetchUsers() {
        long currentTime = System.currentTimeMillis();
        return RestClient.execute("GET", "http://example.com/users&starts_date=" + currentTime, REQUEST_TIMEOUT);
    }

    private int calculateTotalAge() {
        return calculateTotalAge(fetchUsers());
    }

    public int calculateTotalAge(List<User> users) {
        int total = 0;
        for (User user : users) {
            total += calculateAge(user);
        }
        return total;
    }

    private int calculateAge(User user) {
        int userBirthDay = getYearInTimeStr(user.getBirthday());
        int currentYear = Calendar.getInstance().get(Calendar.YEAR);
        return currentYear - userBirthDay;
    }

    /**
     * DEMO METHOD
     * This method will return year from time string
     */
    private static int getYearInTimeStr(String time) {
        return 0;
    }
}

File: iOS/Part1/GetUserInfo/GetUserInfoBetter.swift
import Foundation

class GetUserInfoBetter {
    
    let REQUEST_TIMEOUT = 69
    
    class RestClient {
        /**
         * DEMO METHOD
         * This method make a request to server then receive a list of users
         */
        static func execute(method: String, url: String, timeout: Int) -> [User] {
            return []
        }
    }
    
    // get user from the internet
    private func fetchUsers() -> [User] {
    let currentTime = Date().timeIntervalSince1970
        return RestClient.execute(method: "GET", url: "http://example.com/users&starts_date=\(currentTime)", timeout: REQUEST_TIMEOUT)
    }
    
    private func calculateTotalAge() -> Int {
        return calculateTotalAge(users: fetchUsers())
    }
    
    public func calculateTotalAge(users: [User]) -> Int {
        var total = 0
        for user in users {
            total += calculateAge(user: user)
        }
        return total
    }
    
    private func calculateAge(user: User) -> Int {
        let userBirthDay = getYearInTimeStr(time: user.getBirthday())
        let currentYear = Calendar.current.component(Calendar.Component.year, from: Date())
        return currentYear - userBirthDay
    }
    
    /**
     * DEMO METHOD
     * This method will return year from time string
     */
    private func getYearInTimeStr(time: String) -> Int {
        return 0
    }
}
